mcs/class/corlib/System/TimeZoneInfo.Unity.cs

```diff
@@ -124,44 +122,45 @@ static TimeZoneInfo CreateLocalUnity ()
 		{
 			Int64[] data;
 			string[] names;
+			string daylightDisplayName = null;
 			//Some timezones start in DST on january first and disable it during the summer
 			bool dst_inverted;
+			var adjustmentRulesList = new List<AdjustmentRule> ();
+
+			//the icall only supports years from 1970 through 2037.
+			int firstSupportedDate = 1971;
+			int lastSupportedDate = 2037;
 			int currentYear = DateTime.UtcNow.Year;
+
+			// Find the most recent daylight saving time display name by working backwards
+			bool disableDaylightSavings = true;
+			for (var year = lastSupportedDate; year >= firstSupportedDate; year--) {
+				if (!System.CurrentSystemTimeZone.GetTimeZoneData (year, out data, out names, out dst_inverted))
+					throw new NotSupportedException ("Can't get timezone name.");
+
+				disableDaylightSavings = data[(int)TimeZoneData.AdditionalDaylightOffsetIdx] == 0 || data[(int)TimeZoneData.UtcOffsetIdx] == data[(int)TimeZoneData.AdditionalDaylightOffsetIdx];
+				if (!disableDaylightSavings)
+				{
+					daylightDisplayName = names[(int)TimeZoneNames.DaylightNameIdx];
+					break;
+				}
+			}
+
 			if (!System.CurrentSystemTimeZone.GetTimeZoneData (currentYear, out data, out names, out dst_inverted))
 				throw new NotSupportedException ("Can't get timezone name.");
 
 			var utcOffsetTS = TimeSpan.FromTicks (data[(int)TimeZoneData.UtcOffsetIdx]);
 			char utcOffsetSign = (utcOffsetTS >= TimeSpan.Zero) ? '+' : '-';
 			string displayName = "(GMT" + utcOffsetSign + utcOffsetTS.ToString (@"hh\:mm") + ") Local Time";
 			string standardDisplayName = names[(int)TimeZoneNames.StandardNameIdx];
-			string daylightDisplayName = names[(int)TimeZoneNames.DaylightNameIdx];
+			daylightDisplayName ??= names[(int)TimeZoneNames.DaylightNameIdx];
 
-			var adjustmentRulesList = new List<AdjustmentRule> ();
-			bool disableDaylightSavings = data[(int)TimeZoneData.AdditionalDaylightOffsetIdx] == 0;
 			//If the timezone supports daylight savings time, generate adjustment rules for the timezone
 			if (!disableDaylightSavings) {
-				//the icall only supports years from 1970 through 2037.
-				int firstSupportedDate = 1971;
-				int lastSupportedDate = 2037;
-
-				//first, generate rules from the current year until the last year mktime is guaranteed to supports
-				for (int year = currentYear; year <= lastSupportedDate; year++) {
-					var rulesForCurrentYear = CreateAdjustmentRule (year, out data, out names, standardDisplayName, daylightDisplayName);
-					//breakout if no more rules
-					if (rulesForCurrentYear.Count > 0)
-						adjustmentRulesList.AddRange (rulesForCurrentYear);
-					else
-						break;
-
-				}
-
-				for (int year = currentYear - 1; year >= firstSupportedDate; year--) {
-					var rulesForCurrentYear = CreateAdjustmentRule (year, out data, out names, standardDisplayName, daylightDisplayName);
-					//breakout if no more rules
+				for (int year = firstSupportedDate; year <= lastSupportedDate; year++) {
```

Comment:
Interesting...  any reason you can think of as to why the old code did this in 2 loops (now -> end : now : beginning).

Now you have simplified it?

Reply:
It seems to have been based on faulty assumptions about whether adjustment rules need to be created or not.

The previous behavior was to look for adjustment rules on either side of the current year and stop as soon as a year without adjustment rules was found.

This is one of the causes of the bug, where daylight saving time was simply being ignored for times when a country had gone from not observing DLS to observing DLS and back to not observing, for instance.

We can't early-out from finding adjustment rules, because the presence of _any_ DLS adjustment rules throughout history for a TimeZoneInfo need to be present for the other DLS logic to work correctly.